Sources/Amplitude/Plugins/iOS/UIKitUserInteractions.swift

```diff
@@ -0,0 +1,189 @@
+        let hierarchy: String
+    }
+
+    func eventFromData(with action: Selector) -> UserInteractionEvent {
+        let viewData = extractData(with: action)
+        return UserInteractionEvent(
+            viewController: viewData.viewController,
+            title: viewData.title,
+            accessibilityLabel: viewData.accessibilityLabel,
+            actionMethod: viewData.actionMethod,
+            targetViewClass: viewData.targetViewClass,
+            targetText: viewData.targetText,
+            hierarchy: viewData.hierarchy)
+    }
+
+    func extractData(with action: Selector) -> ViewData {
+        var targetText: String?
+
+        if let button = self as? UIButton {
+            targetText = button.currentTitle
+        } else if let segmentedControl = self as? UISegmentedControl {
+            targetText = segmentedControl.titleForSegment(at: segmentedControl.selectedSegmentIndex)
+        } else {
+            #if !os(tvOS)
+            if #available(iOS 14.0, macCatalyst 14.0, *) {
+                if let colorWell = self as? UIColorWell {
+                    targetText = colorWell.title
+                } else if let `switch` = self as? UISwitch {
+                    targetText = `switch`.title
+                }
+            }
+            #endif
+        }
+
+        let viewController = owningViewController
+        let viewControllerClassName = viewController?.descriptiveTypeName
+        let viewControllerTitle = viewController?.title
+        let targetAccessibilityLabel = self.accessibilityLabel
+        let actionName = NSStringFromSelector(action)
+        let targetViewClassName = self.descriptiveTypeName
+        let viewHierarchy = sequence(first: self, next: { $0.superview })
+            .map { $0.descriptiveTypeName }
+            .joined(separator: UIView.viewHierarchyDelimiter)
+
+        return ViewData(
+            viewController: viewControllerClassName,
+            title: viewControllerTitle,
+            accessibilityLabel: targetAccessibilityLabel,
+            actionMethod: actionName,
+            targetViewClass: targetViewClassName,
+            targetText: targetText,
+            hierarchy: viewHierarchy)
+    }
+}
+
+extension UIResponder {
+    var descriptiveTypeName: String {
+        String(describing: type(of: self))
+    }
+
+    var owningViewController: UIViewController? {
+        return self as? UIViewController ?? next?.owningViewController
+    }
+}
+
+protocol Trackable {
+    func shouldTrack(_ action: Selector, for event: UIEvent?) -> Bool
+}
+
+extension UITextField: Trackable {
+    func shouldTrack(_ action: Selector, for event: UIEvent?) -> Bool {
+        false
+    }
+
+    @objc static func amp_textFieldBeganEditing(_ notification: NSNotification) {
+        guard let textField = notification.object as? UITextField else { return }
+        let userInteractionEvent = textField.eventFromData(with: #selector(UITextField.amp_textFieldBeganEditing))
```

Comment:
Don't record internal selectors.

Reply:
How can we let the customer know that a user ended/began editing a textfield? I was hopping that the selector's name would indicate whether textfield gained/lost focus. Should we create a new property or event for that?

Reply:
The selector name can be very useful to capture, but I'm not sure it will be available as we expand to other capture types, as we are seeing here.

I would think we have some String property like "action"  added to each event, roughly equivalent to the UIControlEvent but in more human understandable terms (ie, click, didEndEditing, etc).